pennylane/tape/jacobian_tape.py

```diff
@@ -224,11 +224,18 @@ def _has_trainable_params(params, diff_methods):
     @staticmethod
     def _flatten_processing_result(g):
         """Flattens the output from processing_fn in parameter shift methods."""
-        if g.dtype is np.dtype("object"):
+        if hasattr(g, "dtype") and g.dtype is np.dtype("object"):
             # object arrays cannot be flattened; must hstack them
+            # check that g has attribute dtype first to allow for
+            # Observables that return other objects that do not have
+            # a dtype
             g = np.hstack(g)
 
-        return g.flatten()
+        if hasattr(g, "flatten"):
+            # flatten only if g supports flattening to allow for
+            # objects other than numpy ndarrays
+            return g.flatten()
+        return g
```

Comment:
Not sure if this is too obfuscating, but we could avoid the if statement by doing this?
```python
getattr(g, "flatten", lambda: g)()
```

Reply:
Too obfuscating for my taste :-)